modules/renter/proto/contract.go

```diff
@@ -96,14 +95,20 @@ type SafeContract struct {
 	headerMu sync.Mutex
 	header   contractHeader
 
-	// merkleRoots are the Merkle roots of each sector stored on the host that
-	// relate to this contract.
-	merkleRoots []crypto.Hash
+	// merkleRoots are the sector roots covered by this contract.
+	merkleRoots *merkleRoots
 
 	// unappliedTxns are the transactions that were written to the WAL but not
 	// applied to the contract file.
 	unappliedTxns []*writeaheadlog.Transaction
 
+	// file is the file of the safe contract that contains the roots. This
+	// file is usually shared with the SafeContract which means multiple
+	// threads could potentially write to the same file. That's why the
+	// SafeContract should never modify the file beyond the
+	// contractHeaderSize and the merkleRoots should never modify data
+	// before that. Both should also use WriteAt and ReadAt instead of
+	// Write and Read.
```

Comment:
can make this safer like so:
```go
type fileSection struct {
	f     *os.File
	start int64
	end   int64
}

func (f *fileSection) WriteAt(p []byte, off int64) (int, error) {
	return f.f.WriteAt(b, f.start+off)
}

// also implement ReadAt, Sync, and Close
```
The tricky part is handling out-of-bounds writes. For the header part of the file, you want to disallow writes past the end. But for the merkle roots part, you need to allow them. Two approaches come to mind:

1) Set `end: -1` for the merkle roots `fileSection`, and check for `f.end == -1` in `WriteAt`.
2) Explicitly extend `f.end` before calling `WriteAt`.

Option 2 seems a bit safer, but I won't make the final call.

The calls to `WriteAt` and `ReadAt` in the merkle roots section also need to updated to use the correct offsets.